stub_uploader/ts_data.py

```diff
@@ -0,0 +1,63 @@
+"""
+Information about typeshed.
+"""
+
+from __future__ import annotations
+import argparse
+from dataclasses import dataclass, fields
+from pathlib import Path
+from tomli import load as toml_load
+from typing import Any
+
+
+REQUIREMENTS = "requirements-tests.txt"
+PYPROJECT = "pyproject.toml"
+
+
+@dataclass
+class TypeshedData:
+    mypy_version: str
+    pyright_version: str
+    pytype_version: str
+
+
+def read_typeshed_data(typeshed_dir: Path) -> TypeshedData:
+    pyproject = _read_pyproject_toml(typeshed_dir / PYPROJECT)
+    requirements = _read_requirements(typeshed_dir / REQUIREMENTS)
+    return TypeshedData(
+        mypy_version=requirements["mypy"],
+        pyright_version=pyproject["tool"]["typeshed"]["pyright_version"],
+        pytype_version=requirements["pytype"],
+    )
+
+
+def _read_pyproject_toml(path: Path) -> dict[str, Any]:
+    with path.open("rb") as f:
+        return toml_load(f)
+
+
+def _read_requirements(path: Path) -> dict[str, str]:
+    requirements = {}
+    with path.open() as f:
+        for line in f:
+            line = line.strip()
+            line = line.split("#")[0]  # strip comments
+            line = line.split(";")[0]  # strip extras
+            if not line.strip():  # skip empty lines
+                continue
+            if "==" in line:
```

Comment:
What if at some point in future someone will (accidentally) write e.g. `mypy>=1.5.1`?

Reply:
Stub uploader already has `packaging` as a requirement, so we could use the `packaging.requirements.Requirement` class to parse each line (once it's been stripped of comments), rather than doing our own parsing. That would simplify this code and also be more robust/better tested.

Reply:
I see that as a feature. Using a non-pinned version in typeshed would make tests non-reproducible. Although maybe we could reorganize requirements-tests.txt a bit and add a comment to that effect?

Reply:
I agree that most of our test dependencies (and definitely our type-checker dependencies) in typeshed should be pinned. But if we want to test that, I feel like we should have a dedicated test for it, rather than having some unrelated code in stub-uploader inexplicably fail when that rule is broken.

Reply:
(We run stub_uploader tests in typeshed, so if you have an integration test that makes sure this works, we should get failures at typeshed PR time)